Second expression in assert statement should be checked as conditional

gantsevdenis opened the pull request:
Fixes https://github.com/python/mypy/issues/7284

ilevkivskyi commented on test-data/unit/check-expressions.test:
```diff
@@ -2303,3 +2303,9 @@ def f(x: any) -> None:  # E: Name 'any' is not defined \
+from typing import Optional, List
+li: Optional[List] = []
+assert li is None, li[0]
+[builtins fixtures/list.pyi]
```
Please keep newline at the end of file.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -2964,14 +2964,16 @@ def visit_operator_assignment_stmt(self,
+        if s.msg is not None:
+            with self.binder.frame_context(can_skip=False, fall_through=1):
+                self.push_type_map(else_map)
+                self.expr_checker.accept(s.msg)
```
It looks like there is a code duplication with `ExpressionChecker.analyze_cond_branch()` why can't you re-use it here? (If there is a reason I am missing please add a comment).

gantsevdenis replied:
I must have missed it!  I ll check that

ilevkivskyi replied:
It looks like there is either a misunderstanding, or you did something wrong in git. I just asked if it is possible to re-use the existing helper here, but you just removed the context manger instead, which is wrong. Namely, if the `else_map` is `None` (i.e. the condition is statically always true), then binder will mark the _current_ frame unreachable.

For example, in this test the missing return statement will be no more caught:
```
[case testAssertNew]
def f() -> int:  # Correct error here on master, no error on your branch.
    x: int
    assert isinstance(x, int), 'Hm...'
[builtins fixtures/isinstance.pyi]
```
Btw, since this was not caught by our test suite, could you please add such test? (Maybe also copy few other basic assert tests adding an assert message to them.)

gantsevdenis commented:
Actually I wasn't sure how to use `analyze_cond_branch()` (I think its 3rd argument, context, is None here), so I went with what I thought, was the safest decision.. turned out to be wrong

So I have added the test that you have suggested, I also added another test with `isinstance` because with first attempt that test would fail while it shoudn't
Thank you for your review time

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -2979,14 +2979,14 @@ def visit_operator_assignment_stmt(self,
+        true_map, else_map = self.find_isinstance_check(s.expr)
+        if s.msg is not None:
+            with self.binder.frame_context(can_skip=False, fall_through=1):
+                self.expr_checker.analyze_cond_branch(else_map, s.msg, None)
```
Hm, sorry for being annoying, but it looks like something is still wrong, now _two_ frames are being pushed, one manually, and one in the helper. Why can't you just use the helper here?

The `None` context is indeed right, because we have no expectations about what should be the assertion "message" (it can be arbitrary object, not just a string).

gantsevdenis replied:
> Why can't you just use the helper here?

Mostly because I misunderstood the binder and its `fall_through` argument
